[inductor] Always realize sigmoid for CPU

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128339

Summary: Currently the cpu backend prefers to always realize exp because it's a heavy op on CPU. For the same reason, we need to realize sigmoid as well. This solves a problem in llama2 inference where exp was repeated in an inner loop for many times.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @chauhang

desertfire committed:
[inductor] Always realize sigmoid for CPU

Summary: Currently the cpu backend prefers to always realize exp because it's a heavy op on CPU. For the same reason, we need to realize sigmoid as well. This solves a problem in llama2 inference where exp was repeated in an inner loop for many times.

[ghstack-poisoned]

eellison reviewed (approved):
looks good - although i wonder if there is more general way we can generate these ? maybe via microbenchmarks ?

not needed for this pr

helloguo reviewed (approved):
Looks good. Now I can see the the sigmoid (which contains exp) is out of the nested loop.

peterbell10 commented on test/inductor/test_cpu_repro.py:
```diff
@@ -3876,6 +3876,48 @@ def fn(arg0_1):
+        opt_fn = torch.compile(fn)
+        _, code = run_and_get_cpp_code(opt_fn, *example_inputs)
+        FileCheck().check_count(
+            "tmp4.store(in_out_ptr0 + static_cast<long>(x0));",
```
This test looks really fragile. Would it make more sense to test the count of `exp` in the graph?

desertfire replied:
The exp will always be in the graph. The difference is if it is inside of a 2-level loop or 1-level loop. I'm open for suggestions on how to check that.

peterbell10 replied:
How about:
```
def fn(x):
    y = x.sigmoid()
    return y + 1, y.sum(-1)
```

Before there are two `exp` calls, one in the pointwise and one in the sum. After there is only one exp call.

jgong5 replied:
Maybe checking the existence of "in_out_ptr" is more robust considering you created a new buffer by realizing sigmoid?

peterbell10 replied:
`in_out_ptr` is an implementation detail that is likely to change soon with the work on `LocalBuffer`. What we really want to test is how many times `exp` is called.

desertfire committed:
Update on "[inductor] Always realize sigmoid for CPU"


Summary: Currently the cpu backend prefers to always realize exp because it's a heavy op on CPU. For the same reason, we need to realize sigmoid as well. This solves a problem in llama2 inference where exp was repeated in an inner loop for many times.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge

desertfire commented:
@pytorchbot merge
